core/src/main/scala/akka/kafka/internal/ActorConsumer.scala

```diff
@@ -0,0 +1,256 @@
+ */
+package akka.kafka.internal
+
+import akka.Done
+import akka.actor.{Actor, ActorRef, ActorSystem, Props, Terminated}
+import akka.event.LoggingReceive
+import akka.kafka.ConsumerSettings
+import akka.kafka.internal.ConsumerStage.CommittableOffsetImpl
+import akka.kafka.scaladsl.Consumer
+import akka.kafka.scaladsl.Consumer.{ClientTopicPartition, CommittableMessage, Control}
+import akka.stream.actor.ActorPublisher
+import akka.stream.actor.ActorPublisherMessage.{Cancel, Request}
+import akka.stream.scaladsl.Source
+import org.apache.kafka.clients.consumer._
+import org.apache.kafka.common.TopicPartition
+
+import scala.annotation.tailrec
+import scala.collection.immutable
+import scala.concurrent.{Future, Promise}
+
+trait MessageBuilder[K, V, MSG] {
+  def createMessage(rec: ConsumerRecord[K, V]): MSG
+}
+
+trait PlainMessageBuilder[K, V] extends MessageBuilder[K, V, ConsumerRecord[K, V]] {
+  self: SourceActor[K, V, ConsumerRecord[K, V]] =>
+
+  override def createMessage(rec: ConsumerRecord[K, V]) = rec
+}
+
+trait CommittableMessageBuilder[K, V] extends MessageBuilder[K, V, CommittableMessage[K, V]] {
+  self: SourceActor[K, V, CommittableMessage[K, V]] =>
+
+  lazy val committer = ActorCommitter(consumer)(context.dispatcher)
+  override def createMessage(rec: ConsumerRecord[K, V]) = {
+    val offset = Consumer.PartitionOffset(
+      ClientTopicPartition(
+        clientId = settings.properties(ConsumerConfig.CLIENT_ID_CONFIG),
+        topic = rec.topic,
+        partition = rec.partition
+      ),
+      offset = rec.offset
+    )
+    Consumer.CommittableMessage(rec.key, rec.value, CommittableOffsetImpl(offset)(committer))
+  }
+}
+
+abstract case class SourceActor[K, V, MSG](settings: ConsumerSettings[K, V])
+    extends ActorPublisher[MSG]
+    with MessageBuilder[K, V, MSG] {
+  var consumer: ActorRef = _
+  var tps = Set.empty[TopicPartition]
+  var buffer: Iterator[ConsumerRecord[K, V]] = Iterator.empty
+
+  override def preStart(): Unit = {
+    super.preStart()
+    consumer = context.actorOf(Props(new KafkaConsumerActor(settings.createKafkaConsumer)))
+    consumer ! KafkaConsumerActor.Subscribe(settings.topics)
+  }
+
+  @tailrec
+  private def pump(): Unit = {
+    if (totalDemand > 0) {
+      if (!buffer.hasNext) {
+        consumer ! KafkaConsumerActor.RequestMessages(tps)
+      }
+      else {
+        val msg = buffer.next()
+        onNext(createMessage(msg))
+        pump()
+      }
+    }
+  }
+
+  override def receive: Receive = working
+
+  def working: Receive = LoggingReceive {
+    case KafkaConsumerActor.Assigned(newTps) =>
+      tps ++= newTps
+      pump()
+    case KafkaConsumerActor.Revoked(newTps) =>
+      tps --= newTps
+      pump()
+    case KafkaConsumerActor.Messages(msgs) =>
+      // do not use simple ++ because of https://issues.scala-lang.org/browse/SI-9766
+      if (buffer.hasNext) {
+        buffer = buffer ++ msgs.asInstanceOf[Iterator[ConsumerRecord[K, V]]]
```

Comment:
What is the difference between `isEmpty` and `hasNext`. For me `hasNext` is fine in `iterator`